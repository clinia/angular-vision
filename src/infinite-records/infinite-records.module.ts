import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgCviInfiniteRecords } from './infinite-records';

@NgModule({
  declarations: [NgCviInfiniteRecords],
  entryComponents: [NgCviInfiniteRecords],
  exports: [NgCviInfiniteRecords],
  imports: [CommonModule],
})
export class NgCviInfiniteRecordsModule {}
