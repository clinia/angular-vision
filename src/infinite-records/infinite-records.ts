import {
  Component,
  ContentChild,
  Input,
  TemplateRef,
  Inject,
  forwardRef,
} from '@angular/core';

import { connectInfiniteRecordsWithInsights } from '@clinia/vision/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgCviVision, Record } from '../vision/vision';
import { noop } from '../utils';

export type InfiniteHitsState = {
  records: Record[];
  results: any;
  isFirstPage: boolean;
  isLastPage: boolean;
  showMore: Function;
  showPrevious: Function;
};

@Component({
  selector: 'cvi-infinite-records',
  template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: state"></ng-container>

      <!-- default rendering if no template specified -->
      <button
        [ngClass]="[
          cx('loadPrevious'),
          this.state.isFirstPage ? cx('loadPrevious', 'disabled') : ''
        ]"
        (click)="showPreviousHandler($event)"
        [disabled]="state.isFirstPage"
        *ngIf="showPrevious && !template"
      >
        {{ showPreviousLabel }}
      </button>

      <div *ngIf="!template">
        <ul [class]="cx('list')">
          <li [class]="cx('item')" *ngFor="let record of state.records">
            {{ record.name }}
          </li>
        </ul>
      </div>

      <button
        [ngClass]="[
          cx('loadMore'),
          this.state.isLastPage ? cx('loadMore', 'disabled') : ''
        ]"
        (click)="showMoreHandler($event)"
        [disabled]="state.isLastPage"
        *ngIf="!template"
      >
        {{ showMoreLabel }}
      </button>
    </div>
  `,
})
export class NgCviInfiniteRecords extends BaseWidget {
  @ContentChild(TemplateRef) public template?: any;

  // rendering options
  @Input() public showPrevious: boolean = false;
  @Input() public showPreviousLabel: string = 'Show previous results';
  @Input() public showMoreLabel: string = 'Show more results';
  @Input() public transformItems?: <U extends Record>(items: Record[]) => U[];

  // inner widget state returned from connector
  public state: InfiniteHitsState = {
    records: [],
    isFirstPage: false,
    isLastPage: false,
    showMore: noop,
    showPrevious: noop,
    results: {},
  };

  constructor(
    @Inject(forwardRef(() => NgCviVision))
    public visionParent: NgCviVision
  ) {
    super('InfiniteRecords');
  }

  ngOnInit() {
    this.createWidget(connectInfiniteRecordsWithInsights, {
      transformItems: this.transformItems,
    });
    super.ngOnInit();
  }

  public showMoreHandler(event: MouseEvent) {
    event.preventDefault();
    this.state.showMore();
  }

  public showPreviousHandler(event: MouseEvent) {
    event.preventDefault();
    this.state.showPrevious();
  }

  updateState = (state, isFirstRendering: boolean) => {
    if (isFirstRendering) return;
    this.state = state;
  };
}
