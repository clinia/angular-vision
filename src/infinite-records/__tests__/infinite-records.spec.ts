import { connectInfiniteRecordsWithInsights } from '@clinia/vision/es/connectors';
import { createRenderer } from '../../../test/test-renderer';
import { NgCviInfiniteRecords } from '../infinite-records';

const defaultState = {
  records: [
    { id: '1', name: 'foo', description: 'foo' },
    { id: '2', name: 'bar', description: 'bar' },
    { id: '3', name: 'foobar', description: 'foobar' },
    { id: '4', name: 'barfoo', description: 'barfoo' },
  ],
  showMore: jest.fn(),
  isLastPage: false,
};

describe('InfiniteHits', () => {
  it('renders markup without state', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-infinite-records></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-infinite-records></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should create widget with connectInfiniteRecordsWithInsights and pass instance options', () => {
    const createWidget = jest.spyOn(
      NgCviInfiniteRecords.prototype,
      'createWidget'
    );

    const transformItems = jest.fn(x => x);
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-infinite-records [transformItems]="transformItems"></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
      methods: { transformItems },
    });

    render({});

    expect(createWidget).toHaveBeenCalledWith(
      connectInfiniteRecordsWithInsights,
      {
        transformItems,
      }
    );
    createWidget.mockRestore();
  });

  it('should call `showMore()` on button click', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-infinite-records></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const showMore = jest.fn();
    const fixture = render({ showMore });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadMore'
    );
    button.click();

    expect(button.disabled).toEqual(false);
    expect(showMore).toHaveBeenCalled();
  });

  it('should disable `showMore` button', () => {
    const render = createRenderer({
      defaultState,
      template: '<cvi-infinite-records></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });

    const showMore = jest.fn();
    const fixture = render({ showMore, isLastPage: true });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadMore'
    );
    button.click();

    expect(button.disabled).toEqual(true);
    expect(showMore).not.toHaveBeenCalled();
    expect(fixture).toMatchSnapshot();
  });

  it('should display `showPrevious()` button', () => {
    const showPrevious = jest.fn();
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-infinite-records [showPrevious]=true></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({ showPrevious });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadPrevious'
    );

    expect(button).toBeTruthy();
  });

  it('should display `showPrevious()` button with custom label', () => {
    const showPrevious = jest.fn();
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-infinite-records [showPrevious]=true showPreviousLabel="Load previous"></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({ showPrevious });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadPrevious'
    );

    expect(button.innerHTML).toMatchInlineSnapshot(`" Load previous "`);
  });

  it('should call `showPrevious()` on button click', () => {
    const showPrevious = jest.fn();
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-infinite-records [showPrevious]=true></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({ showPrevious });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadPrevious'
    );
    button.click();

    expect(button.disabled).toEqual(false);
    expect(showPrevious).toHaveBeenCalledTimes(1);
  });

  it('should disable `showPrevious` button on first page', () => {
    const showPrevious = jest.fn();
    const render = createRenderer({
      defaultState,
      template:
        '<cvi-infinite-records [showPrevious]=true></cvi-infinite-records>',
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({ showPrevious, isFirstPage: true });

    const button = fixture.debugElement.nativeElement.querySelector(
      '.cvi-InfiniteRecords-loadPrevious'
    );

    expect(button.className).toMatchInlineSnapshot(
      `"cvi-InfiniteRecords-loadPrevious cvi-InfiniteRecords-loadPrevious--disabled"`
    );
    expect(button.disabled).toEqual(true);
    expect(showPrevious).toHaveBeenCalledTimes(0);
    expect(fixture).toMatchSnapshot();
  });

  it('should render with custom templates', () => {
    const render = createRenderer({
      defaultState,
      template: `
        <cvi-infinite-records>
          <ng-template
            let-records="records"
            let-showMore="showMore"
            let-showPrevious="showPrevious"
            let-isFirstPage="isFirstPage"
            let-isLastPage="isLastPage"
          >
            <button (click)="showPrevious()" [disabled]="isFirstPage">Load previous</button>
            <div *ngFor="let record of records">
              <strong>{{record.name}}</strong>
            </div>
            <button (click)="showMore()" [disabled]="isLastPage">Load more</button>
          </ng-template>
        </cvi-infinite-records>
      `,
      TestedWidget: NgCviInfiniteRecords,
    });
    const fixture = render({});

    expect(fixture).toMatchSnapshot();
  });

  it('should allow calling insightsClient', () => {
    const render = createRenderer({
      defaultState,
      template: `
        <cvi-infinite-records>
          <ng-template
            let-records="records"
            let-insights="insights"
          >
            <ul>
              <li *ngFor="let record of records">
                <button
                  id="add-to-cart-{{record.id}}"
                  (click)="insights('clickedRecordIDsAfterSearch', { eventName: 'Add to cart', ids: [record.id] })">

                </button>
              </li>
            </ul>
          </ng-template>
        </cvi-infinite-records>
      `,
      TestedWidget: NgCviInfiniteRecords,
    });
    const insights = jest.fn();
    const fixture = render({ insights });
    expect(fixture).toMatchSnapshot();

    const button = fixture.debugElement.nativeElement.querySelector(
      '#add-to-cart-2'
    );
    button.click();

    expect(insights).toHaveBeenCalledWith('clickedRecordIDsAfterSearch', {
      eventName: 'Add to cart',
      ids: ['2'],
    });
  });
});
