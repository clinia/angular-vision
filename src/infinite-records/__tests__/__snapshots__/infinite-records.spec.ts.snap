// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteHits renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        
        
        <div>
          <ul
            class="cvi-InfiniteRecords-list"
          >
            
            <li
              class="cvi-InfiniteRecords-item"
            >
               foo 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               bar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               foobar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               barfoo 
            </li>
          </ul>
        </div>
        
        <button
          class="cvi-InfiniteRecords-loadMore"
        >
           Show more results 
        </button>
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;

exports[`InfiniteHits renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        
        
        <div>
          <ul
            class="cvi-InfiniteRecords-list"
          >
            
          </ul>
        </div>
        
        <button
          class="cvi-InfiniteRecords-loadMore"
        >
           Show more results 
        </button>
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;

exports[`InfiniteHits should allow calling insightsClient 1`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        <ul>
          
          <li>
            <button
              id="add-to-cart-1"
            />
          </li>
          <li>
            <button
              id="add-to-cart-2"
            />
          </li>
          <li>
            <button
              id="add-to-cart-3"
            />
          </li>
          <li>
            <button
              id="add-to-cart-4"
            />
          </li>
        </ul>
        
        
        
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;

exports[`InfiniteHits should disable \`showMore\` button 1`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        
        
        <div>
          <ul
            class="cvi-InfiniteRecords-list"
          >
            
            <li
              class="cvi-InfiniteRecords-item"
            >
               foo 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               bar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               foobar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               barfoo 
            </li>
          </ul>
        </div>
        
        <button
          class="cvi-InfiniteRecords-loadMore cvi-InfiniteRecords-loadMore--disabled"
          disabled=""
        >
           Show more results 
        </button>
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;

exports[`InfiniteHits should disable \`showPrevious\` button on first page 2`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        
        <button
          class="cvi-InfiniteRecords-loadPrevious cvi-InfiniteRecords-loadPrevious--disabled"
          disabled=""
        >
           Show previous results 
        </button>
        
        <div>
          <ul
            class="cvi-InfiniteRecords-list"
          >
            
            <li
              class="cvi-InfiniteRecords-item"
            >
               foo 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               bar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               foobar 
            </li>
            <li
              class="cvi-InfiniteRecords-item"
            >
               barfoo 
            </li>
          </ul>
        </div>
        
        <button
          class="cvi-InfiniteRecords-loadMore"
        >
           Show more results 
        </button>
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;

exports[`InfiniteHits should render with custom templates 1`] = `
<ng-component
  testedWidget={[Function NgCviInfiniteRecords]}
>
  <cvi-vision>
    <cvi-infinite-records>
      <div
        class="cvi-InfiniteRecords"
      >
        
        <button>
          Load previous
        </button>
        
        <div>
          <strong>
            foo
          </strong>
        </div>
        <div>
          <strong>
            bar
          </strong>
        </div>
        <div>
          <strong>
            foobar
          </strong>
        </div>
        <div>
          <strong>
            barfoo
          </strong>
        </div>
        <button>
          Load more
        </button>
        
        
        
      </div>
    </cvi-infinite-records>
  </cvi-vision>
</ng-component>
`;
